### ISSUE TITLE
In-context OD

### ISSUE THREAD

**Issue by cutz-j** [planemanner/IncontextGLaMM#4]
#### 일정
- ~모델입력 & 모듈 수정(\~8/15)~
- ~loss 수정 (\~8/22)~
- ~loss 수정~ & ~validation코드수정~ & ~pairedtransform 오류 수정~ & ~학습진행~ (~9/1)
- ~deepspped 적용해서 학습 후 결과 확인 (~9/8)~
- objectness score 추가하여 재학습 (~9/21)
- paper 초안 (~9/28)

#### 고민
- class 처리 -> 지금 단계에서는 생략
- masking 방식 -> 실험
- loss 구성 -> l1 으로 하고 안되면 찾아보자

#### [정리중] Related Works
- [ICLR22] DAB-DETR: DYNAMIC ANCHOR BOXES ARE BETTER QUERIES FOR DETR
  - 기존 DETR구조에서 learnable quries -> reference points로 단순히 query_embedding을 decoder에 주었던 DETR과 달리 learnable anchors를 도입하고 decoder에서도 이를 활용할 수 있도록 변경
- [WACV24] Self-Supervised Learning for Visual Relationship Detection through Masked Bounding Box Reconstruction
  - bbox에 해당하는 object pooling embedding(Faster-RCNN)을 마스킹하여 self-sup learning
- [CVPR21] Instance Localization for Self-supervised Detection Pretraining
  - ROI에 해당하는 object feature를 background와 합성하여 contrastive learning
- [CVPR22] DN-DETR: Accelerate DETR Training by Introducing Query DeNoising
  - denising queries 적용: gt bbox/label에 noise(random scaling, prob_noise) 를 추가하여 indicator로 추가
    - 일반 부분은 bbox_loss 계산 = hungarian matching
    -  denoising query: 매칭없이 GT를 직접 구성하도록 focal, l1, GIoU 
  - 일반 query와 함께 denosiing query를 추가한 형태로 detr 학습 convergence & 속도 단축에 기여함
  - denoising query를 모델에 입력으로 쓰는 방식을 차용하면 좋을 것으로 보임
 - rcnn -> adaptation 비교

#### [완] Paired-Dataset: (image1, image2), (bbox1, bbox2), (cls1, cls2) -> height concat + bbox 통합
- Dataset: coco -> vision(industrial domain)
  - /aidata_02/owod/data/coco_2017
  - /aidata_02/owod/data/vision
- datasets.__init__ -> "coco_icl" --> build_coco_icl
- pair_idx 구현
- paired_transform 구현
  - 800 width 고정
- _combine_images -> concat (dim=1): (c, h, w) -> 아래로 붙인 형태
- _combine bbox -> bbox 수정 필요 -> bbox를 추가한 형태 (N1+N2, 1+4)
- ~class idx 수정 필요 -> concat~, **class 우선 제외**
- PairedRandomSizeCrop 내에 crop_w, crop_h randint size 에러 발생
  - RandomResize -> RandomResize -> RandomSizeCrop -. RandomResize 메커니즘

#### [완] Masking 모듈 추가: Masking target bbox -> (N1+N2, 4)
- Masking 방식 고민 (Masked Modeling for Self-supervised Representation Learning on Vision and Beyond)
  - raw coordinate masking [num_box, (cx, cy, w, h)]
  - seggpt patch embedding masking
    - segmap(c=3) -> conv2d(learnable, c=768) -> y*mask_token(learnable)*boolean_mask(c=3)
  - seggpt pixel-wise masking -> 실험
  - bert token masking: continuous variable -> discrete variable 정보손실 가능성
  - Beitv2 patch embedding masking: 학습이 잘될까
  - MBBR masking: roi pooling feature masking
- util/masking_generator.py - raw coordinate masking: num_box * coord의 masking ratio 만큼 masking
  - masking value = -1 -> sampling location 구할 때 위험할 수 있음
  - mask_tensor -> decoder 전달 연산 필요가 없는건가?
  - nq만큼 padding할 수 있도록 추가 (nb+@=nq, 4)

#### [완] 모델 입력 & BBOX DAB 모듈 수정
- paired_img, tgt 입력
- maksed_boxes tgt 입력 변경 -> query_embeds
  - tgt_embed = (bs, nq, 256)
  - ref_point = (bs, nq, 4)
- classification module 정리
  - dab_deformable_detr.py 내 forward -> logits 부분 삭제

#### [완] Loss 수정
- 기존 Loss
  - bbox l1 loss -> masking 하지 않은 부분을 제외하고 평균을 구하도록 수정 필요
  - bbox giou loss -> prediction과 gt 계산 가능
  - cardinality loss: 예측 cls (no obj 제외) 개수와 tgt 개수 l1 -> 계산불가능
  - label sigmoid focal loss -> 불가능
- SetCriterion 수정
  - forward 수정
    - num_boxes -> boxes length
  - get_loss 추가 (mask_bbox)
- matcher 수정
  - matcher.py 내에 HungarianMatcherBbox 클래스 추가 (bbox만 계산하도록 변경 & masking 부분만 계산하도록 적용) 
- loss_mask_bbox 추가
    - targets['masked_bbox'] -> masking 제외 (mask_tensor)
    - target_masks 정의 -> loss_bbox * target_masks -> sum() -> / num_boxes
- loss 계산 process
  - SetCriterion(num_classes, matcher, weight_dict, losses, focal_alpha)
  - aux_loss: 이전 레이어의 output도 loss 계산 -> hungarian matching 불안정성때문에 적용
  - matcher -> HungarianMatcher
  - bs, num_queries -> 비용행렬구성 (cost_class, cost_bbox, cost_giou) -> matching
    - 부분 masking을 고려한 cost 코드 수정
    - masking loss 계산시에는 masking 제외부분은 loss 계산 X
    - masking 구역의 prediction과 gt bbox 비교
    - Hungarian matching은 gt bbox와 cost 행렬을 계산하되 mask 부분은 가리도록 해야함 -> masking을 가릴 필요가 없음
      - 결국 prediction과 gt간의 가장 가까운 거리를 찾아서 1:1 매칭을 시켜주기 위함일뿐
      -  이 과정에서 masking이 되지 않은 부분이 큰 역할을 할 뿐 masking 부분을 가릴 필요가 없다
    - 결국은 Hungarian matching은 query들의 prediction을 gt개수만큼 1:1 매칭시켜주는 역할 -> loss 계산이 가능해짐

#### [완] train
  - 일부 class_embed 제거
- 학습 pair 구성시에 동일 클래스로 묶는게 나을지 아니면 전체 랜덤(현재): 전체랜덤 -> 안되면 묶자
- LLM 추가 -> 클래스 prediction task 추가
- context 정의가 새로 필요할지도 -> idea... 추후 대미래

#### [완] validation
- data_loader_val 수정
  - build_dataset -> image_set='val'
  - coco_icl내에서 image_set 받을 수 있도록 수정, tgt1 idx와 tgt2 idx를 알 수 있도록 tgt_idx 추가
  - masking_generator 에서 train/val에 따라 random masking / tgt2 masking을 처리할 수 있도록 수정
- evaluate 코드 수정
 - model inference 이후에 outputs, targets 수정 -> tgt1_idx, tgt2_idx에 따라서 수정
  - tgt1, pred1 부분을 아예 제거하고 평가할수있도록 변경
  - outputs -> tgt1_idx에 따라 pred_boxes crop
  - targets -> tgt1_idx에 따라 boxes,  mask_tensor, masked_boxes 수정 

#### debug
- classification 모듈 혹은 일부 미사용 모듈로 grad 에러 발생
  - 모든 class_embed 주석처리
- paired transform 과정에서 crop_h 계산과정에서 empty range for randrange() (384, 323, -61)    
  - crop_h random idx를 고르는 과정에서 concat image를 염두할 수 있도록 img1.height+img2.height ~ self.max_size 로 변경
  - 에러가 지속발생하여 randomresizecrop transform 우선 제외
- masking generator에서 masking idx를 고르는 random choice 함수에서 대상 idx 리스트가 아예 없어서 에러나는 경우 존재
  - 확인 불가 -> randomresizecrop transform 문제로 보임

### todo
- 학습방식 - hard case training
- masking 방식 -> bbox 단위 -> bbox가 있어도되고 없어도되는 상황이 매우 문제
- pairing 방식 -> 동일 class
- num_query 감소 (300 --> 50)
- inference시에 img1-tgt1도 제대로 prediction 못하는 부분
  - underfitting 가능성
  - masking 제외부분은 학습에서 제외한 부분 -> 전체 학습을 하지만 masking 부분에 가중치를 두도록 변경
- objectness의 경우 결국 학습데이터셋에서의 matching에 의해 구분될텐데, domain shift에서 제대로 구현이 될 수 있는지
- center collapse 해결하기 위해 진행한 방안
  - matcher -> unmasking bbox만 matching 되도록 조정, cost_obj = 0
  - loss -> masking bbox 직접 매칭 (l1, giou loss만 적용) / unmasking bbox는 matching 이후 연결
  - masking generator -> masking ratio = 0.5로 조정
  - reference point (gt input) 에서 masking token (-10)이 직접 모델 입력에 쓰이지 않도록 rand 값으로 변경
  - curriculum learning 적용: 학습 초기에 2 epoch까지는 unmask bbox만 학습하도록 변경, 이후 masking box 직접매칭을 학습하도록 점진적으로 증가

#### Deepspeed
- deepspeed 적용가능하도록 install & code setup
- ms_deform_attn function이 float16, bfloat16을 지원하지 않음
  - 해당 function만 float32로 돌아갈 수 있도록 추가

#### inference
- query_num = 300개 만큼의 prediction을 유효한 prediction으로 줄일 수 있는 방법이 필요함
- vision dataset loader 코드 필요

#### DAB 모듈 흐름
- DAB-deformable-transformer 흐름
  - DAB_DETR (dab_deformable_detr.py)
    - tgt_embed weight(nq, 256), refpoint_embed weight(nq, 4) concat -> query_embeds(nq, 300) -> transformer
  - DeformableTransformer (deformable_transform.py) 
    -  query_embeds split -> referecne_points (nq, 4) sigmoid = init_reference_out (ref)
    - tgt unsqueeze (bs, nq, 256) (decoder 입력) -> decoder
  - DeformableTransformerDecoder
    - reference_points repeat (bs, nq, 4) -> reference_points_input = product by src_valid_ratios (bs, nq, 4, 4)? -> decoder_layer
    - raw_query_pos = ref_point_head (2~3 layer FFN) (bs, nq, 256) 
    - output = tgt -> pos_scale = query_scale(2 layer ffpn) (bs, nq, 256)
    - query_pos = pos_scale * raw_query_pos -> decoder layer
  - DeformableTransformerDecoderLayer
    - q = k = tgt + query_pos -> MHAttention(q, k, tgt) = tgt2
    - tgt + dropout(tgt2) -> layernorm = tgt
    - tgt2 = MSDefomAtten(tgt+query_pos, reference_points, encoder_output,..) -> layernorm -> ffn
  - MSDeformableAttn (ms_deform_attn.py)
    - query = tgt+query_pos(weight)
    - input_flatten=encoder_output -> value_proj(input_flatten) (linear) = value + input_padding_mask
    - sampling_offsets = sampling_offesets(query) (linear)
    - attention_weights = attention_weights(query) (linear) -> softmax
    - sampling_locations = reference_points + sampling_offesets -> MSDeformAttnFunction(value, sampling location) -> output_proj = output

#### 결과 정리
- 학습셋에서의 center collapse 가 매우 크게 발생 
  - 원인은 무엇일까?
    - dab deformable detr 구조상 reference point의 주변 포인트를 바탕으로 query_embed에 반영이 되게끔 decoding을 진행함
    - reference point에서 masking이 들어갈 경우 random point로 들어가게 되는데 이 point 근처로만 bbox가 prediction 되는것으로 보임.
    - 만약 deformable attention module에서 masking reference point를 보지 않도록 학습한다면? 궁극적으로 icl inference에서 원하는 bbox inference가 가능한가? --> 매우 애매함 (query_embed에서 학습때 prediction하지 않는 masking bbox prediction을 할 수 없을 듯함)
    - 아예 input에서 합친다면? seggpt는 image와 segmentation의 차원이 비슷하지만, image와 bbox의 차원은 매우 imbalance함
